### PR TITLE
feat: picture component and refactor to attribute selectors

### PR DESCRIPTION
**Pull request by frederickfogerty**
This PR introduces picture and source components to add `<picture>` support to this library. In the making I realised that since Angular adds wrapper elements to every component that uses a custom component name (e.g. `ix-img`), this wasn't going to work for the picture component. I did a lot of research and made the hard decision to switch to attribute selectors. This makes the usage a little bit more confusing, but I think this is a good change in the long run. As a side benefit, developers no longer have to use `htmlAttributes` to set html attributes on the underlying elements, and can just access them directly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -315,18 +325,53 @@ The last way to implement lazy loading is to use an event listener. This is not
...
 
+#### Picture Support
+
+With the picture element, images can be directed to have different crops and sizes based on the browser dimensions, or any media query.
+
+It is recommended to check out our [introduction blog post about how to best use picture and the imgix API](https://docs.imgix.com/tutorials/using-imgixPicture-element) for some great tips and examples!
+
+##### Fluid mode
+
+By default, the srcset values generated for are fluid-width srcsets, will respond with the actual size of your picture element page, as a fluid-width image would.
```
Comment (sherwinski) [inline README.md:349]: This sentence doesn't read very clearly, might just to be slightly reworded.
```suggestion
By default, the srcset values generated are fluid-width srcsets. This allows them to respond with the actual size of your picture element on the page.
```

File: projects/imgix-ng/src/common/index.ts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+
+
```
Comment (sherwinski) [inline projects/imgix-ng/src/common/index.ts:2]: Did something get missed here?

File: projects/imgix-ng/src/lib/imgix.component.ts [hunk 3]
```diff
@@ -68,35 +64,28 @@ export class ImgixComponent implements AfterViewChecked {
...
-
-    if (!(config.includeLibraryParam === false)) {
-      (client as any).settings.libraryParam = `ng-${VERSION}`;
-    }
+  @Input() htmlAttributes?: Record<string, string>;
 
-    return client;
+  constructor(
+    @Inject(ImgixConfigService) private config: ImgixConfig,
+    private elementRef: ElementRef<HTMLImageElement>,
```
Comment (sherwinski) [inline projects/imgix-ng/src/lib/imgix.component.ts:71]: Does this get used?

File: projects/imgix-ng/src/lib/picture.spec.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+        const srcsets = srcset.split(',');
+        const getQuality = (srcset) => srcset.match(/q=\d*/)[0].slice(2);
+
+        const firstQuality = getQuality(srcsets[0]);
+        const lastQuality = getQuality(srcsets[srcsets.length - 1]);
+
+        expect(firstQuality).not.toBe(lastQuality);
+      });
+      it('the srcset should have a variable quality when disableVariableQuality attribute set', async () => {
```
Comment (sherwinski) [inline projects/imgix-ng/src/lib/picture.spec.ts:98]:
```suggestion
      it('the srcset should not have a variable quality when disableVariableQuality attribute set', async () => {
```

File: projects/imgix-ng/src/test/url-assert.ts [hunk 5]
```diff
@@ -1,6 +1,9 @@
 export const stringMatchingFixedSrcSet = (width: number) =>
   jasmine.stringMatching(new RegExp(`.*w=${width}&.*dpr=1.* 1x,`));
 
+export const stringMatchingFixedSrcSetHeight = (heigh: number) =>
+  jasmine.stringMatching(new RegExp(`.*h=${heigh}&.*dpr=1.* 1x,`));
```
Comment (sherwinski) [inline projects/imgix-ng/src/test/url-assert.ts:5]:
```suggestion
export const stringMatchingFixedSrcSetHeight = (height: number) =>
  jasmine.stringMatching(new RegExp(`.*h=${height}&.*dpr=1.* 1x,`));
```

**Review by sherwinski — APPROVED** [review]
Overall, looks good to me 👍 

More general question: Should we still refer to the exported elements as `Component`? I'm wondering if that could be potentially confusing.

**Comment by frederickfogerty** [comment]
Thanks for the review @sherwinski. I'll address these issues in a subsequent PR.

> More general question: Should we still refer to the exported elements as Component? I'm wondering if that could be potentially confusing.

So, technically, the exported items are still "components" - they're just using an attribute selector rather than a custom tag name. Furthermore, since adding `ixImg` to a tag converts that tag into a component (i.e. the attribute values are passed to that component), whereas for a directive it is more restricted, I think it's accurate to keep calling it a component.

File: projects/imgix-ng/src/lib/imgix.component.ts (hunk 3 above)
Comment (frederickfogerty) [reply projects/imgix-ng/src/lib/imgix.component.ts:71]: Yep, in `setOtherAttributes()` and `setSrcAndSrcsetAttributes()`

File: projects/imgix-ng/src/lib/picture.spec.ts (hunk 4 above)
Comment (frederickfogerty) [reply projects/imgix-ng/src/lib/picture.spec.ts:98]: Thanks!

**Comment by sherwinski** [comment]
Ah right because there's still a `@Component` decorator. Makes sense 👍

**Pull request merged**